File: .config/ags/components/clock.ts
interface OutputRecord {
	hour: string;
	minute: string;
	second: string;
	tooltip: string;
}

const record = Variable<OutputRecord>({
	hour: '',
	minute: '',
	second: '',
	tooltip: '',
});

Utils.subprocess(
	'/home/minhnbnt/.config/eww/bar/bin/calendar',
	(output) => (record.value = JSON.parse(output)),
);

export default function Clock() {
	return Widget.Box({
		orientation: 1,
		tooltip_text: record.bind().as((record) => record.tooltip),
		children: [
			Widget.Label({
				label: record.bind().as((record) => record.hour),
			}),
			Widget.Label({
				label: record.bind().as((record) => record.minute),
			}),
			Widget.Label({
				label: record.bind().as((record) => record.second),
			}),
		],
	});
}

File: .config/ignis/widgets/clock.py
from collections.abc import Callable
from datetime import datetime

from ignis.utils.poll import Poll
from ignis.widgets import Widget

_clock = Poll(1000, lambda _: datetime.now())


def transform(func: Callable[[datetime], str]):
    return _clock.bind("output", func)


def clock() -> Widget.Box:
    return Widget.Box(
        vertical=True,
        halign="center",
        css_classes=("clock",),
        child=(
            Widget.Label(label=transform(lambda t: f"{t.hour:02d}")),
            Widget.Label(label=transform(lambda t: f"{t.minute:02d}")),
            Widget.Label(label=transform(lambda t: f"{t.second:02d}")),
        ),
        tooltip_text=transform(lambda t: t.strftime("%A, %B %d %Y %H:%M:%S")),
    )
